### PR TITLE
[3.2] SHiP Fix forking behavior

### PR DESCRIPTION
**Pull request by heifner**
`state_history_plugin` was not sending updated blocks when `nodeos` forked. Enhanced the `ship_streamer_test` to generate a fork and verify SHiP sends new blocks on forks.

### CODE REVIEW

File: tests/ship_streamer_test.py [hunk 1]
```diff
@@ -57,28 +59,99 @@
     cluster.killall(allInstances=killAll)
     cluster.cleanup()
     Print("Stand up cluster")
+
+
+    # ***   setup topogrophy   ***
+
+    # "bridge" shape connects defprocera through defproducerc (3 in node0) to each other and defproduceru (1 in node01)
```
Comment (linh2931) [inline tests/ship_streamer_test.py:66]: node01 should be node1

File: tests/ship_streamer_test.py [hunk 2]
```diff
@@ -9,39 +9,41 @@
...
-#   is configured with the state_history_plugin.  An instance
-#   of node will be started with ship_streamer to exercise
-#   the SHiP API.
+# This test sets up 2 producing nodes and one "bridge" node using test_control_api_plugin.
+#   One producing node has 3 of the elected producers and the other has 1 of the elected producers.
+#   All the producers are named in alphabetical order, so that the 3 producers, in the one production node, are
+#       scheduled first, followed by the 1 producer in the other producer node. Each producing node is only connected
+#       to the other producing node via the "bridge" node.
+#   The bridge node has the test_control_api_plugin, that the test uses to kill
+#       the "bridge" node to generate a fork.
```
Comment (linh2931) [inline tests/ship_streamer_test.py:23]: The comments are only about network setup. You can add something about ship testing.

File: tests/ship_streamer_test.py [hunk 3]
```diff
@@ -57,28 +59,99 @@
...
+    Print("Creating wallet \"%s\"." % (testWalletName))
+    testWallet=walletMgr.create(testWalletName, [cluster.eosioAccount,accounts[0],accounts[1],accounts[2],accounts[3],accounts[4],accounts[5]])
+
+    for _, account in cluster.defProducerAccounts.items():
+        walletMgr.importKey(account, testWallet, ignoreDupKeyWarning=True)
+
+    for i in range(0, totalNodes):
+        node=cluster.getNode(i)
+        node.producers=Cluster.parseProducers(i)
+        numProducers=len(node.producers)
```
Comment (linh2931) [inline tests/ship_streamer_test.py:116]: `numProducers` not used.

File: tests/ship_streamer_test.py [hunk 4]
```diff
@@ -57,28 +59,99 @@
...
+        for prod in node.producers:
+            prodName = cluster.defProducerAccounts[prod].name
+            if prodName == "defproducera" or prodName == "defproducerb" or prodName == "defproducerc" or prodName == "defproduceru":
+                Print("Register producer %s" % cluster.defProducerAccounts[prod].name)
```
Comment (linh2931) [inline tests/ship_streamer_test.py:120]: Why not `cluster.defProducerAccounts[prod].name)` just be `prodName`?

File: tests/ship_streamer_test.py [hunk 5]
```diff
@@ -57,28 +59,99 @@
...
+                trans=node.regproducer(cluster.defProducerAccounts[prod], "http::/mysite.com", 0, waitForTransBlock=False, exitOnError=True)
```
Comment (linh2931) [inline tests/ship_streamer_test.py:121]: Only one / in `http::/mysite.com`. I don't know what this trans is for.

File: tests/ship_streamer_test.py [hunk 6]
```diff
@@ -57,28 +59,99 @@
...
+        nonProdNode.transferFunds(cluster.eosioAccount, account, transferAmount, "test transfer", waitForTransBlock=True)
+        trans=nonProdNode.delegatebw(account, 20000000.0000, 20000000.0000, waitForTransBlock=False, exitOnError=True)
+
+    # ***   vote using accounts   ***
+
+    cluster.waitOnClusterSync(blockAdvancing=3)
     start_block_num = shipNode.getBlockNum()
-    end_block_num = start_block_num + args.num_blocks
+
+    # vote a,b,c  u
```
Comment (linh2931) [inline tests/ship_streamer_test.py:137]: Why space between c and u?

File: tests/ship_streamer_test.py [hunk 7]
```diff
@@ -112,7 +215,7 @@
         data = json.load(outFile)
         block_num = start_block_num
         for i in data:
-            print(i)
+            #print(i)
```
Comment (linh2931) [inline tests/ship_streamer_test.py:218]: Remove this commented out line.

File: plugins/state_history_plugin/state_history_plugin.cpp [hunk 8]
```diff
@@ -351,6 +351,7 @@ struct state_history_plugin_impl : std::enable_shared_from_this<state_history_pl
```
Comment (linh2931) [inline plugins/state_history_plugin/state_history_plugin.cpp:351]: I am curious about the `if (!current_request || !current_request->max_messages_in_flight)` above. Why return when `max_messages_in_flight` is false?

File: tests/CMakeLists.txt [hunk 9]
```diff
@@ -98,12 +98,12 @@ target_link_libraries(ship_client abieos Boost::program_options Boost::system Th
 add_executable(ship_streamer ship_streamer.cpp)
 target_link_libraries(ship_streamer abieos Boost::program_options Boost::system Threads::Threads)
 
-add_test(NAME ship_test COMMAND tests/ship_test.py -v --num-clients 1 --num-requests 5000 --clean-run --dump-error-detail WORKING_DIRECTORY ${CMAKE_BINARY_DIR})
+add_test(NAME ship_test COMMAND tests/ship_test.py -v --num-clients 3 --num-requests 5000 --clean-run --dump-error-detail WORKING_DIRECTORY ${CMAKE_BINARY_DIR})
```
[the remark this answers is not in the document]
Comment (heifner) [reply tests/CMakeLists.txt:101]: I do think we need >1. I'll change to 10.

File: tests/ship_streamer_test.py (hunk 5 above)
Comment (heifner) [reply tests/ship_streamer_test.py]: This registers the producer, the website can be any string, nothing verifies it.

File: tests/ship_streamer_test.py (hunk 2 above)
Comment (heifner) [reply tests/ship_streamer_test.py:23]: Added comments about the ship part of the test.

File: plugins/state_history_plugin/state_history_plugin.cpp (hunk 8 above)
Comment (linh2931) [reply plugins/state_history_plugin/state_history_plugin.cpp:351]: So why the second part not `current_request->max_messages_in_flight` but `!current_request->max_messages_in_flight`?

Comment (heifner) [reply plugins/state_history_plugin/state_history_plugin.cpp:351]: Ah, look at line 342, it counts down.

File: tests/ship_streamer.cpp [hunk 10]
```diff
@@ -128,12 +130,41 @@ int main(int argc, char* argv[]) {
          eosio::check(result_doucment[1]["head"].IsObject(),                                 "'head' is not an object");
          eosio::check(result_doucment[1]["head"].HasMember("block_num"),                     "'head' does not contain 'block_num'");
          eosio::check(result_doucment[1]["head"]["block_num"].IsUint(),                      "'head.block_num' isn't a number");
+         eosio::check(result_doucment[1]["head"].HasMember("block_id"),                      "'head' does not contain 'block_id'");
+         eosio::check(result_doucment[1]["head"]["block_id"].IsString(),                     "'head.block_num' isn't a string");
```
Comment (oschwaldp-oci) [inline tests/ship_streamer.cpp:134]: Should this be `result_document`?

File: tests/ship_streamer_test.py [hunk 11]
```diff
@@ -57,28 +63,98 @@
...
+    accounts[0].name="testeraaaaaa"
+    accounts[1].name="tester111111" # needed for voting
+    accounts[2].name="tester222222" # needed for voting
+    accounts[3].name="tester333333" # needed for voting
+    accounts[4].name="tester444444" # needed for voting
+    accounts[5].name="tester555555" # needed for voting
+
+    testWalletName="test"
+
+    Print("Creating wallet \"%s\"." % (testWalletName))
```
Comment (oschwaldp-oci) [inline tests/ship_streamer_test.py:111]: Would prefer f-strings in new code. There are several throughout, but I'll only comment here on this one so if you choose not to update, there is only one comment to close.

File: tests/ship_test.py [hunk 12]
```diff
@@ -132,7 +132,7 @@
             try:
                 statuses = json.loads(" ".join(lines))
             except json.decoder.JSONDecodeError as er:
-                Utils.errorExit("javascript client output was malformed in %s. Exception: %s" % (shipClientErrorFile, er))
+                Utils.errorExit("ship_client output was malformed in %s. Exception: %s" % (shipClientErrorFile, er))
```
Comment (oschwaldp-oci) [inline tests/ship_test.py:135]: Would prefer f-strings when updating. I'll only comment here on this one so if you choose not to update, there is only one comment to close.

File: tests/ship_streamer.cpp (hunk 10 above)
Comment (heifner) [reply tests/ship_streamer.cpp:134]: lol, Took me awhile to figure out what you were saying. I'll fix the spelling.

File: tests/ship_streamer_test.py (hunk 11 above)
Comment (heifner) [reply tests/ship_streamer_test.py]: Changed all to use f-strings

File: tests/ship_test.py (hunk 12 above)
Comment (heifner) [reply tests/ship_test.py:135]: Since only minor changes in this python script, did not update to using f-string.

**Review by spoonincode — APPROVED** [review]

**Pull request merged**